lib/experimental/fonts-library/class-wp-font-family.php

```diff
@@ -0,0 +1,416 @@
+	 *
+	 * @param string $src The path of the font asset file to delete.
+	 * @return bool Whether the file was deleted.
+	 */
+	static private function delete_asset( $src ) {
+		$filename  = basename( $src );
+		$file_path = path_join( WP_FONTS_DIR, $filename );
+
+		if ( ! file_exists( $file_path ) ) {
+			return false;
+		}
+		wp_delete_file( $file_path );
+
+		// If the file still exists after trying to delete it, return false.
+		if ( file_exists( $file_path ) ) {
+			return false;
+		}
+
+		// return true if the file was deleted.
+		return true;
+	}
+
+	/**
+	 * Deletes all font face asset files associated with a given font face.
+	 *
+	 * @param array $font_face The font face array containing the 'src' attribute with the file path(s) to be deleted.
+	 */
+	static private function delete_font_face_assets( $font_face ) {
+		$srcs = ! empty( $font_face['src'] ) && is_array( $font_face['src'] )
+			? $font_face['src']
+			: array( $font_face['src'] );
+		foreach ( $srcs as $src ) {
+			$was_asset_removed = self::delete_asset( $src );
+			if ( ! $was_asset_removed ) {
+				// Bail if any of the assets could not be removed.
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Downloads a font asset.
+	 *
+	 * Downloads a font asset from a specified source URL and saves it to the font directory.
+	 *
+	 * @param string $src The source URL of the font asset to be downloaded.
+	 * @param string $filename The filename to save the downloaded font asset as.
+	 * @return string|bool The relative path to the downloaded font asset. False if the download failed.
+	 */
+	private function download_asset( $src, $filename ) {
+		$file_path = path_join( WP_FONTS_DIR, $filename );
+
+		// Checks if the file to be downloaded has a font mime type.
+		if ( ! WP_Font_Family_Utils::has_font_mime_type( $file_path ) ) {
+			return false;
+		}
+
+		// Include file with download_url() if function doesn't exist.
+		if ( ! function_exists( 'download_url' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/file.php';
+		}
+
+		// Downloads the font asset or returns false.
+		$temp_file = download_url( $src );
+		if ( is_wp_error( $temp_file ) ) {
+			return false;
+		}
+
+		// Moves the file to the fonts directory or return false.
+		$renamed_file = rename( $temp_file, $file_path );
+		// Cleans the temp file.
+		@unlink( $temp_file );
+
+		if ( ! $renamed_file ) {
+			return false;
+		}
+
+		// Returns the relative path to the downloaded font asset to be used as font face src.
+		return WP_Fonts_Library::get_relative_fonts_path() . $filename;
+	}
+
+	/**
+	 * Move an uploaded font face asset from temp folder to the wp fonts directory.
+	 *
+	 * This is used when uploading local fonts.
+	 *
+	 * @param array $font_face Font face to download.
+	 * @param array $file Uploaded file.
+	 * @return array New font face with all assets downloaded and referenced in the font face definition.
+	 */
+	private function move_font_face_asset( $font_face, $file ) {
```

Comment:
Hmm, I wonder ... Core has a `_wp_handle_upload()` that handles the PHP uploads into WordPress. It does stuff such as:

* Sanitizes the file anems
* Checks extensions for mime type
* Moves the the file into the appropriate directory
* and has an error handler too

While this function targets the `uploads` directory, I'm thinking the internals of the function should be used and then targeted to the `fonts` directory.

Pinging the Filesystem and Upgrades components maintainer @costdev as well as @azaozz for their expertise,

Reply:
The Uploads directory functionality has years and years and years of know-how and testing built into it. It wraps around PHP functionality for different server environments and provides a variety of hooks for extenders to customize it.

Though these are targeting the `uploads` directory, I'm wondering if these can be used in an efficient and maintainable way to point to the `fonts` directory instead of `uploads` directory.

Why?

To reuse it and avoid duplicating the code while gaining the benefits of all the know-how that has been built into it over the years.

Reply:
> The Uploads directory functionality has years and years and years of know-how and testing built into it.
 Yep, good idea to reuse these functions, no need to "reinvent" all that and have to fix all the edge cases again :)

Reply:
@azaozz @hellofromtonya It seems like it is not currently possible to write a file into a folder different to the `uploads` directory using `_wp_handle_upload`.

Considering [this comment](https://github.com/WordPress/gutenberg/pull/52704#discussion_r1271167455), I'm not sure if you still think we should you the uploads directory or if we can add `wp-content/fonts`. Could you provide more input/context, please?

In both cases, adding `wp-content/fonts` and writing `wp-content/uploads/fonts`: how could indicate `_wp_handle_upload` as the path we want to write?

Reply:
> In both cases, adding `wp-content/fonts` and writing `wp-content/uploads/fonts`: how could indicate to `_wp_handle_upload` the path we want to write?

Theoretically, we could hook [upload_dir](https://developer.wordpress.org/reference/hooks/upload_dir/):
```php
add_filter( 'upload_dir', array( $this, 'set_upload_dir' ) );
// $file_was_moved = wp_handle_upload(
remove_filter( 'upload_dir', array( $this, 'set_upload_dir' ) );
// if ( isset( $file_was_moved['file'] ) ) { <- Untested.

public function set_upload_dir( $uploads ) {
	// Modify $uploads or return a new array:
	return array(
		'path'    => '',
		'url'     => '',
		'subdir'  => '',
		'basedir' => '',
		'baseurl' => '',
		'error'   => false,
	);
}
```

Reply:
Implementing the use of `wp_handle_upload` here: wp_handle_upload